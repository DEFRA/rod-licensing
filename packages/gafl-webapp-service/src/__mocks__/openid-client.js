export const mockSettings = {
  ENV: {
    OIDC_ENDPOINT: 'https://oauth-endpoint/',
    OIDC_REDIRECT_HOST: 'https://localhost:3143',
    OIDC_CLIENT_ID: 'client-id',
    OIDC_CLIENT_SECRET: 'client-secret',
    OIDC_SESSION_COOKIE_NAME: 'oidc-cookie-name',
    OIDC_SESSION_COOKIE_PASSWORD: 'open-sesame',
    OIDC_REQUIRE_DYNAMICS_ROLE: 'Telesales-role'
  },
  TEST_AUTH_ENDPOINT: 'https://oauth-endpoint/token-endpoint',
  TEST_AUTH_ISSUER: 'https://oauth-endpoint/issuer',
  TEST_NONCE: 'test_nonce',
  TEST_STATE: 'test_state',
  TEST_OID: 'test-azure-object-id',
  TEST_NAME: 'user fullname',
  TEST_EMAIL: 'user@example.com',
  TEST_EXP: 9999999999,
  TEST_POST_AUTH_REDIRECT: '/test/redirect'
}

const OpenIdClient = jest.genMockFromModule('openid-client')
OpenIdClient.Issuer = {
  discover: jest.fn(async () => ({
    issuer: mockSettings.TEST_AUTH_ISSUER,
    metadata: {},
    Client: mockOidcClientConstructor
  }))
}
OpenIdClient.generators.state = jest.fn(() => mockSettings.TEST_STATE)
OpenIdClient.generators.nonce = jest.fn(() => mockSettings.TEST_NONCE)

export const setupEnvironment = () => {
  Object.assign(process.env, mockSettings.ENV)
}

export const mockOidcClient = {
  authorizationUrl: jest.fn(() => mockSettings.TEST_AUTH_ENDPOINT),
  callback: jest.fn(() => ({
    claims: jest.fn(() => ({
      oid: mockSettings.TEST_OID,
      name: mockSettings.TEST_NAME,
      email: mockSettings.TEST_EMAIL,
      exp: mockSettings.TEST_EXP
    }))
  }))
}

export const mockOidcClientConstructor = jest.fn(() => mockOidcClient)

export const mockCache = {
  get: jest.fn(async () => {
    return {
      nonce: mockSettings.TEST_NONCE,
      state: mockSettings.TEST_STATE,
      postAuthRedirect: mockSettings.TEST_POST_AUTH_REDIRECT
    }
  }),
  set: jest.fn(),
  clear: jest.fn()
}
export const mockServer = {
  cache: jest.fn(() => mockCache),
  auth: {
    strategy: jest.fn(),
    default: jest.fn()
  }
}

export default OpenIdClient
