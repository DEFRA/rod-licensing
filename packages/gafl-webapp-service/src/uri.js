/*
 * Page locations, templates
 */
export const LICENCE_LENGTH = { uri: '/buy/licence-length', page: 'licence-length' }
export const LICENCE_TYPE = { uri: '/buy/licence-type', page: 'licence-type' }
export const LICENCE_TO_START = { uri: '/buy/start-kind', page: 'licence-to-start' }
export const LICENCE_START_TIME = { uri: '/buy/start-time', page: 'licence-start-time' }

export const NO_LICENCE_REQUIRED = { uri: '/buy/no-licence-required', page: 'no-licence-required' }
export const DISABILITY_CONCESSION = { uri: '/buy/disability-concession', page: 'disability-concession' }
export const DATE_OF_BIRTH = { uri: '/buy/date-of-birth', page: 'date-of-birth' }

export const NAME = { uri: '/buy/name', page: 'name' }
export const ADDRESS_LOOKUP = { uri: '/buy/find-address', page: 'address-lookup' }
export const ADDRESS_SELECT = { uri: '/buy/select-address', page: 'address-select' }
export const ADDRESS_ENTRY = { uri: '/buy/address', page: 'address-entry' }
export const CONTACT = { uri: '/buy/contact', page: 'contact' }
export const NEWSLETTER = { uri: '/buy/newsletter', page: 'newsletter' }

export const LICENCE_FULFILMENT = { uri: '/buy/fulfilment', page: 'licence-fulfilment' }
export const LICENCE_CONFIRMATION_METHOD = { uri: '/buy/confirmation-method', page: 'licence-confirmation-method' }

export const CONTACT_SUMMARY = { uri: '/buy/contact-summary', page: 'contact-summary' }
export const LICENCE_SUMMARY = { uri: '/buy/licence-summary', page: 'licence-summary' }

export const PAYMENT_CANCELLED = { uri: '/buy/payment-cancelled', page: 'payment-cancelled' }
export const PAYMENT_FAILED = { uri: '/buy/payment-failed', page: 'payment-failed' }

export const TERMS_AND_CONDITIONS = { uri: '/buy/terms-conditions', page: 'terms-and-conditions' }
export const ORDER_COMPLETE = { uri: '/buy/order-complete', page: 'order-complete' }
export const LICENCE_INFORMATION = { uri: '/buy/licence-information', page: 'licence-information' }
export const ATTRIBUTION = { uri: '/buy/attribution' }

export const OIDC_SIGNIN = { uri: '/oidc/signin' }
export const OIDC_ROLE_REQUIRED = { uri: '/oidc/role-required', page: 'role-required' }
export const OIDC_ACCOUNT_DISABLED = { uri: '/oidc/account-disabled', page: 'account-disabled' }

/**
 * Renewals pages
 * @type {{uri: string}}
 */
export const RENEWAL_PUBLIC = { uri: '/renew/{referenceNumber?}' }
export const IDENTIFY = { uri: '/buy/renew/identify', page: 'identify' }
export const RENEWAL_INACTIVE = { uri: '/buy/renew/inactive', page: 'renewal-inactive' }
export const AUTHENTICATE = { uri: '/buy/renew/authenticate' }
export const RENEWAL_START_DATE = { uri: '/buy/renew/renewal-start-date', page: 'renewal-start-date' }
export const RENEWAL_START_VALIDATE = { uri: '/buy/renew/renewal-start-date/validate' }

export const CONTROLLER = { uri: '/buy' }
export const NEW_TRANSACTION = { uri: '/buy/new' }
export const ADD_PERMISSION = { uri: '/buy/add' }
export const AGREED = { uri: '/buy/agreed' }

export const CLIENT_ERROR = { uri: '/buy/client-error', page: 'client-error' }
export const SERVER_ERROR = { uri: '/buy/server-error', page: 'server-error' }

/**
 * These are informational static pages
 */
export const COOKIES = { uri: '/guidance/cookies', page: 'cookies' }
export const ACCESSIBILITY_STATEMENT = { uri: '/guidance/accessibility-statement', page: 'accessibility-statement' }
export const PRIVACY_POLICY = { uri: '/guidance/privacy-policy', page: 'privacy-policy' }
export const REFUND_POLICY = { uri: '/guidance/refund-policy', page: 'refund-policy' }
export const OS_TERMS = { uri: '/guidance/os-terms', page: 'os-terms' }

/**
 * These are inserted at runtime by the test framework but the session manager needs to know about them
 */
export const TEST_STATUS = { uri: '/buy/status' }
export const TEST_TRANSACTION = { uri: '/buy/transaction' }
export const GET_PRICING_TYPES = { uri: '/buy/get-pricing/types' }
export const GET_PRICING_LENGTHS = { uri: '/buy/get-pricing/lengths' }

export const FRESHWATER_FISING_RULES = { uri: 'https://www.gov.uk/freshwater-rod-fishing-rules' }
export const LOCAL_BYELAWS = { uri: 'https://www.gov.uk/government/collections/local-fishing-byelaws' }
