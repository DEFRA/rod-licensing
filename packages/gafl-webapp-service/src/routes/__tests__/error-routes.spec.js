import { setupEnvironment } from '../../__mocks__/openid-client.js'
import { CLIENT_ERROR } from '../../uri.js'
import errorRoutes from '../error-routes'

let TestUtils = null

describe('Error route handlers', () => {
  beforeAll(async () => {
    jest.isolateModules(() => {
      process.env.ERROR_PAGE_ROUTE = 'true'
      setupEnvironment()
      TestUtils = require('../../__mocks__/test-utils-system.js')
      TestUtils.start(() => {})
    })
  })
  beforeEach(jest.clearAllMocks)

  // Stop application after running the test case
  afterAll(async () => {
    TestUtils.stop(() => {})
  })

  it('redirects to the client error page when client error is thrown', async () => {
    const data = await TestUtils.server.inject({
      method: 'GET',
      url: '/buy/client-error'
    })
    expect(data.statusCode).toBe(302)
  })

  describe('CLIENT_ERROR route', () => {
    it('should pass the catalog and language to the view if it is present', async () => {
      const request = getMockRequest()
      const mockToolkit = getMockToolkit()
      const clientError = errorRoutes[0].handler
      await clientError(request, mockToolkit)
      expect(mockToolkit.view).toBeCalledWith(
        CLIENT_ERROR.page,
        expect.objectContaining({
          mssgs: [],
          altLang: []
        })
      )
    })

    describe.each([
      [true, { payment_id: 'abc123', href: 'gov.pay.url' }],
      [true, { payment_id: 'def456', href: 'gov-pay-url' }],
      [false, {}]
    ])('includes correct data when paymentInProgress is %p', (paymentInProgress, payment) => {
      it(`includes paymentInProgress flag with value set to ${paymentInProgress} to correspond to presence of payment_id ${payment.payment_id}`, async () => {
        const request = getMockRequest(payment)
        const mockToolkit = getMockToolkit()
        const clientError = errorRoutes[0].handler
        await clientError(request, mockToolkit)
        expect(mockToolkit.view).toHaveBeenCalledWith(
          expect.any(String),
          expect.objectContaining({
            paymentInProgress
          })
        )
      })

      it(`${paymentInProgress ? 'includes' : 'excludes'} govpay url`, async () => {
        const request = getMockRequest(payment)
        const mockToolkit = getMockToolkit()
        const clientError = errorRoutes[0].handler
        await clientError(request, mockToolkit)
        if (paymentInProgress) {
          expect(mockToolkit.view).toHaveBeenCalledWith(
            expect.any(String),
            expect.objectContaining({
              uri: expect.objectContaining({
                payment: payment.href
              })
            })
          )
        } else {
          expect(mockToolkit.view).not.toHaveBeenCalledWith(
            expect.any(String),
            expect.objectContaining({
              uri: expect.objectContaining({
                payment: payment.href
              })
            })
          )
        }
      })
    })
  })
})

const getMockToolkit = (view = jest.fn(() => ({ code: () => {} }))) => ({
  view
})

const getMockRequest = (payment = {}) => ({
  cache: () => ({
    helpers: {
      transaction: {
        get: () =>
          Promise.resolve({
            payment
          })
      }
    }
  }),
  headers: {},
  i18n: {
    getCatalog: () => [],
    getLocales: () => []
  },
  response: {
    isBoom: true,
    output: {
      statusCode: 400
    }
  },
  url: 'url',
  path: 'path',
  query: 'query',
  params: 'params',
  payload: 'payload',
  state: 'state',
  method: 'method'
})
