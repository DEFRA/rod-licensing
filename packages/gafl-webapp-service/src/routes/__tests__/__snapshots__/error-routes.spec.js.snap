// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error route handlers error route CLIENT_ERROR handler handler should return correct values 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "client-error",
      Object {
        "altLang": Array [],
        "clientError": "payload",
        "mssgs": Array [],
        "paymentInProgress": false,
        "uri": Object {
          "new": "/buy/new",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "code": [Function],
      },
    },
  ],
}
`;

exports[`Error route handlers error route SERVER_ERROR handler handler should return correct values 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "server-error",
      Object {
        "altLang": Array [],
        "mssgs": Array [],
        "serverError": "payload",
        "uri": Object {
          "agreed": "/buy/agreed",
          "new": "/buy/new",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "code": [Function],
      },
    },
  ],
}
`;

exports[`Error route handlers error route has a return value with a method of GET and path of /buy/client-error and /buy/server-error 1`] = `
Array [
  Object {
    "handler": [Function],
    "method": Array [
      "GET",
    ],
    "path": "/buy/client-error",
  },
  Object {
    "handler": [Function],
    "method": Array [
      "GET",
    ],
    "path": "/buy/server-error",
  },
]
`;
